File: src/pipeline/chunk.ts
import { z } from "zod";
import type { NormalizedDevEvent, SessionChunk } from "../adapters/types.js";
import { callHaiku } from "../llm/client.js";

// ── Constants ────────────────────────────────────────────────────────

const PAUSE_THRESHOLD_MS = 5 * 60 * 1000; // 5 minutes
const FILE_SHIFT_THRESHOLD = 0.7; // 70% new files
const MIN_PREVIOUS_FILES = 3; // need enough file context to detect a real shift
const MIN_CURRENT_FILES = 1; // at least one file needed
const MIN_CHUNK_SIZE = 8; // merge tiny chunks into neighbors
const SIZE_CAP = 80;
const OVERLAP = 3;
const SHORT_SESSION_THRESHOLD = 10;

// Ignore files outside the project — skill docs, tmp files, etc.
const IGNORED_FILE_PATTERNS = [
  /\/\.claude\//,
  /\/node_modules\//,
  /^\/tmp\//,
  /^\/private\/tmp\//,
];

// ── Topic-Shift Detection (Haiku) ────────────────────────────────────
// Semantic classification — never regex. We ask Haiku which user (intent)
// messages mark an explicit pivot to a NEW task/topic. The result is a set
// of event IDs that `chunkSession` consumes as a deterministic signal, so
// chunking itself stays pure and the LLM call is opt-in (skipped on dry-run).

const TopicShiftSchema = z
  .object({
    shifts: z
      .array(
        z
          .object({
            eventId: z.string(),
            isTopicShift: z.boolean().optional().default(false),
          })
          .passthrough(),
      )
      .optional()
      .default([]),
  })
  .passthrough();

/** Max user messages sent to Haiku in a single classification call. */
const MAX_TOPIC_SHIFT_CANDIDATES = 250;
/** Truncate each message to keep the payload bounded. */
const TOPIC_SHIFT_TEXT_CAP = 280;

/**
 * Classify which user messages mark an explicit topic/task shift, using Haiku
 * structured output (replaces the old regex phrase matching).
 *
 * Returns a set of event IDs that begin a new topic. On any failure — no API
 * key, network error, short session, no candidates — returns an empty set so
 * the deterministic split signals (pauses, file-cluster shifts) still apply
 * and the parent pipeline never fails because of this enrichment.
 */
export async function detectTopicShifts(
  events: NormalizedDevEvent[],
): Promise<Set<string>> {
  const empty = new Set<string>();

  // Short sessions are never split (mirrors chunkSession's short-circuit).
  if (events.length < SHORT_SESSION_THRESHOLD) return empty;

  const candidates = events
    .filter((e) => e.category === "intent")
    .slice(0, MAX_TOPIC_SHIFT_CANDIDATES);
  if (candidates.length < 2) return empty;

  try {
    const { system, user } = buildTopicShiftPrompt(candidates);
    const result = await callHaiku(system, user, TopicShiftSchema);
    const ids = new Set<string>();
    for (const s of result.shifts) {
      if (s.isTopicShift) ids.add(s.eventId);
    }
    return ids;
  } catch (err) {
    process.stderr.write(
      `  ⚠ Topic-shift detection failed (${err instanceof Error ? err.message : String(err)}). Continuing without topic-shift splits.\n`,
    );
    return empty;
  }
}

function buildTopicShiftPrompt(candidates: NormalizedDevEvent[]): {
  system: string;
  user: string;
} {
  const system = `You analyze the developer's messages within a single AI-assisted coding session and identify which messages mark an EXPLICIT shift to a NEW topic or task.

A topic shift is when the developer deliberately pivots away from what was just being worked on to start a different piece of work — for example: "now let's work on X", "moving on to Y", "switch to Z", "ok, next: ...", "different thing now".

The following are NOT topic shifts: continuations of the current task, clarifications, answers to a question the assistant asked, bug reports about the in-progress work, approvals/rejections ("yes", "no", "looks good"), and refinements of the same goal.

Be conservative — only flag a message when the developer is clearly starting new work.

Respond with ONLY a JSON object of the form:
{ "shifts": [ { "eventId": "<id>", "isTopicShift": true|false }, ... ] }
Include an entry for every message id you were given.`;

  const lines = candidates.map(
    (e) =>
      `- id: ${e.id}\n  message: ${(e.content.detail || e.content.summary || "").slice(0, TOPIC_SHIFT_TEXT_CAP)}`,
  );

  const user = `Developer messages, in order:

${lines.join("\n")}

For each message id, decide whether it marks an explicit shift to a new topic/task. Return JSON only.`;

  return { system, user };
}

// ── Main Entry Point ─────────────────────────────────────────────────

/**
 * Split a session's normalized events into bounded reasoning windows (chunks).
 * Deterministic — no LLM calls. Topic-shift split signals are supplied as a
 * precomputed set of event IDs (see `detectTopicShifts`); when omitted, only
 * pause and file-cluster-shift signals are used.
 */
export function chunkSession(
  events: NormalizedDevEvent[],
  sessionId: string,
  topicShiftEventIds: Set<string> = new Set(),
): SessionChunk[] {
  if (events.length === 0) return [];

  // Short sessions: one chunk, no splitting
  if (events.length < SHORT_SESSION_THRESHOLD) {
    return [buildChunk(events, sessionId, 0, [])];
  }

  // Find all split points
  const splitIndices = findSplitPoints(events, topicShiftEventIds);

  // Build chunks from split points
  const rawChunks = splitAtIndices(events, splitIndices);

  // Merge tiny chunks into neighbors
  const mergedChunks = mergeTinyChunks(rawChunks);

  // Apply size cap to oversized chunks
  const cappedChunks = mergedChunks.flatMap((chunk) => applySizeCap(chunk));

  // Build SessionChunk objects with overlap
  return buildChunksWithOverlap(cappedChunks, sessionId);
}

// ── Split Point Detection ────────────────────────────────────────────

/**
 * Find indices where the event list should be split.
 * A split index N means: chunk boundary between events[N-1] and events[N].
 */
function findSplitPoints(
  events: NormalizedDevEvent[],
  topicShiftEventIds: Set<string>,
): number[] {
  const splits: number[] = [];
  let lastSplit = 0; // Enforce minimum gap between splits

  for (let i = 1; i < events.length; i++) {
    // Don't split if we're too close to the last split
    if (i - lastSplit < MIN_CHUNK_SIZE) continue;

    // Priority 1: Large pauses (>5 min gap)
    const gap =
      new Date(events[i].timestamp).getTime() -
      new Date(events[i - 1].timestamp).getTime();
    if (gap > PAUSE_THRESHOLD_MS) {
      splits.push(i);
      lastSplit = i;
      continue;
    }

    // Priority 2: File cluster shift
    if (hasFileClusterShift(events, i)) {
      splits.push(i);
      lastSplit = i;
      continue;
    }

    // Priority 3: Explicit topic shift in user messages (Haiku-classified)
    if (hasTopicShift(events[i], topicShiftEventIds)) {
      splits.push(i);
      lastSplit = i;
    }
  }

  return splits;
}

/**
 * Check if the forward window of files at index i represents a cluster shift
 * compared to the preceding window.
 *
 * Compares files from events[index..index+LOOKAHEAD] against
 * files from events[index-LOOKBACK..index-1].
 * Triggers when >70% of the forward window's files are new.
 */
function hasFileClusterShift(
  events: NormalizedDevEvent[],
  index: number,
): boolean {
  const LOOKAHEAD = 5;
  const LOOKBACK = 15;

  // Collect files from forward window (current + next few events)
  const forwardEnd = Math.min(events.length, index + LOOKAHEAD);
  const forwardFiles = new Set<string>();
  for (let j = index; j < forwardEnd; j++) {
    for (const f of getProjectFiles(events[j])) {
      forwardFiles.add(f);
    }
  }
  if (forwardFiles.size < MIN_CURRENT_FILES) return false;

  // Collect files from the preceding window
  const lookbackStart = Math.max(0, index - LOOKBACK);
  const previousFiles = new Set<string>();
  for (let j = lookbackStart; j < index; j++) {
    for (const f of getProjectFiles(events[j])) {
      previousFiles.add(f);
    }
  }

  if (previousFiles.size < MIN_PREVIOUS_FILES) return false;

  // Check what percentage of forward files are new
  const forwardArr = [...forwardFiles];
  const newFiles = forwardArr.filter((f) => !previousFiles.has(f));
  return newFiles.length / forwardArr.length > FILE_SHIFT_THRESHOLD;
}

/**
 * Check if an event was classified as an explicit topic shift by Haiku.
 */
function hasTopicShift(
  event: NormalizedDevEvent,
  topicShiftEventIds: Set<string>,
): boolean {
  if (event.category !== "intent") return false;
  return topicShiftEventIds.has(event.id);
}

// ── Merge Tiny Chunks ────────────────────────────────────────────────

/**
 * Merge chunks smaller than MIN_CHUNK_SIZE into their preceding neighbor.
 * If the first chunk is tiny, merge it into the next one.
 */
function mergeTinyChunks(
  chunks: NormalizedDevEvent[][],
): NormalizedDevEvent[][] {
  if (chunks.length <= 1) return chunks;

  const result: NormalizedDevEvent[][] = [chunks[0]];

  for (let i = 1; i < chunks.length; i++) {
    if (chunks[i].length < MIN_CHUNK_SIZE) {
      // Merge into the previous chunk
      result[result.length - 1] = [
        ...result[result.length - 1],
        ...chunks[i],
      ];
    } else {
      result.push(chunks[i]);
    }
  }

  // If the first chunk ended up tiny (from being the original first), merge forward
  if (result.length > 1 && result[0].length < MIN_CHUNK_SIZE) {
    result[1] = [...result[0], ...result[1]];
    result.shift();
  }

  return result;
}

// ── Chunk Building ───────────────────────────────────────────────────

/**
 * Split events array at the given indices into sub-arrays.
 */
function splitAtIndices(
  events: NormalizedDevEvent[],
  indices: number[],
): NormalizedDevEvent[][] {
  if (indices.length === 0) return [events];

  const chunks: NormalizedDevEvent[][] = [];
  let start = 0;

  for (const idx of indices) {
    if (idx > start) {
      chunks.push(events.slice(start, idx));
    }
    start = idx;
  }

  // Remaining events after the last split
  if (start < events.length) {
    chunks.push(events.slice(start));
  }

  return chunks;
}

/**
 * If a chunk exceeds SIZE_CAP, split it at natural boundaries.
 */
function applySizeCap(events: NormalizedDevEvent[]): NormalizedDevEvent[][] {
  if (events.length <= SIZE_CAP) return [events];

  const result: NormalizedDevEvent[][] = [];
  let start = 0;

  while (start < events.length) {
    if (events.length - start <= SIZE_CAP) {
      result.push(events.slice(start));
      break;
    }

    // Find the nearest natural boundary around SIZE_CAP
    const target = start + SIZE_CAP;
    const splitAt = findNaturalBoundary(events, target);
    result.push(events.slice(start, splitAt));
    start = splitAt;
  }

  return result;
}

/**
 * Find a natural boundary near the target index.
 * Prefer splitting after a tool_result or before a conversation_turn.
 * Search within +/- 10 events of target.
 */
function findNaturalBoundary(
  events: NormalizedDevEvent[],
  target: number,
): number {
  const searchRadius = 10;
  const lo = Math.max(1, target - searchRadius);
  const hi = Math.min(events.length, target + searchRadius);

  // Prefer: after tool_result (split at i+1)
  for (let i = target; i >= lo; i--) {
    if (events[i - 1]?.category === "result") return i;
  }
  for (let i = target + 1; i < hi; i++) {
    if (events[i - 1]?.category === "result") return i;
  }

  // Fallback: before conversation_turn
  for (let i = target; i >= lo; i--) {
    if (events[i]?.category === "intent") return i;
  }
  for (let i = target + 1; i < hi; i++) {
    if (events[i]?.category === "intent") return i;
  }

  // Last resort: split at target
  return Math.min(target, events.length);
}

/**
 * Build final SessionChunk objects with overlap between consecutive chunks.
 */
function buildChunksWithOverlap(
  rawChunks: NormalizedDevEvent[][],
  sessionId: string,
): SessionChunk[] {
  const result: SessionChunk[] = [];

  for (let i = 0; i < rawChunks.length; i++) {
    const ownedEvents = rawChunks[i];
    // Prepend overlap from previous chunk (last OVERLAP events)
    const overlapEvents =
      i > 0 ? rawChunks[i - 1].slice(-OVERLAP) : [];

    result.push(buildChunk(ownedEvents, sessionId, i, overlapEvents));
  }

  return result;
}

/**
 * Build a single SessionChunk from owned events and optional overlap events.
 */
function buildChunk(
  ownedEvents: NormalizedDevEvent[],
  sessionId: string,
  chunkIndex: number,
  overlapEvents: NormalizedDevEvent[],
): SessionChunk {
  const allEvents = [...overlapEvents, ...ownedEvents];

  const startCausalOrder = ownedEvents[0].causalOrder;
  const endCausalOrder = ownedEvents[ownedEvents.length - 1].causalOrder;

  return {
    id: `${sessionId}-chunk-${chunkIndex}`,
    sessionId,
    chunkIndex,
    events: allEvents,
    topicHint: deriveTopicHint(ownedEvents),
    filesInScope: deriveFilesInScope(allEvents),
    eventRange: [startCausalOrder, endCausalOrder],
  };
}

// ── Topic Hint & Files ───────────────────────────────────────────────

/**
 * Derive a topic hint from the chunk's owned events.
 * Heuristic: most frequent directory/filename from filesAffected,
 * falling back to first user message summary.
 */
function deriveTopicHint(events: NormalizedDevEvent[]): string {
  // Count file path frequency
  const freq = new Map<string, number>();
  for (const event of events) {
    for (const file of getFilesAffected(event)) {
      freq.set(file, (freq.get(file) ?? 0) + 1);
    }
  }

  if (freq.size > 0) {
    // Return the most frequent file path
    let best = "";
    let bestCount = 0;
    for (const [file, count] of freq) {
      if (count > bestCount) {
        best = file;
        bestCount = count;
      }
    }
    return best;
  }

  // Fallback: first user message summary
  const userEvent = events.find((e) => e.category === "intent");
  if (userEvent) {
    return userEvent.content.summary;
  }

  return "unknown";
}

/**
 * Collect all unique files from events' filesAffected.
 */
function deriveFilesInScope(events: NormalizedDevEvent[]): string[] {
  const files = new Set<string>();
  for (const event of events) {
    for (const f of getFilesAffected(event)) {
      files.add(f);
    }
  }
  return Array.from(files);
}

/**
 * Get filesAffected from an event, defaulting to empty array.
 */
function getFilesAffected(event: NormalizedDevEvent): string[] {
  return event.content.filesAffected ?? [];
}

/**
 * Get only project-relevant files (filter out skill docs, tmp, node_modules).
 * Used for file cluster shift detection — we don't want reading a skill doc
 * to trigger a chunk split.
 */
function getProjectFiles(event: NormalizedDevEvent): string[] {
  return getFilesAffected(event).filter(
    (f) => !IGNORED_FILE_PATTERNS.some((p) => p.test(f)),
  );
}

File: backend/quire/ingest/chunk.py
"""Split a session's normalized events into bounded reasoning windows (chunks).

Port of journal/src/pipeline/chunk.ts — exact algorithmic equivalence with the
DETERMINISTIC path (no LLM calls). In the TS pipeline, topic-shift detection
uses Haiku; the dry-run path calls chunkSession with an empty topicShiftEventIds
set, so the Python port's default is also an empty set.

Constants (must match TS exactly):
    PAUSE_THRESHOLD_MS = 5 * 60 * 1000   (5 minutes)
    FILE_SHIFT_THRESHOLD = 0.7            (70% new files)
    MIN_PREVIOUS_FILES = 3
    MIN_CURRENT_FILES = 1
    MIN_CHUNK_SIZE = 8
    SIZE_CAP = 80
    OVERLAP = 3
    SHORT_SESSION_THRESHOLD = 10
"""

from __future__ import annotations

import re
from typing import Optional

from .models import NormalizedDevEvent, SessionChunk
from .sittings import _parse_ts

# ── Constants (must match TS chunk.ts exactly) ────────────────────────
PAUSE_THRESHOLD_MS = 5 * 60 * 1000   # 5 minutes
FILE_SHIFT_THRESHOLD = 0.7            # 70% new files trigger a cluster shift
MIN_PREVIOUS_FILES = 3
MIN_CURRENT_FILES = 1
MIN_CHUNK_SIZE = 8
SIZE_CAP = 80
OVERLAP = 3
SHORT_SESSION_THRESHOLD = 10

# Files outside the project are ignored for file-cluster shift detection
_IGNORED_FILE_PATTERNS = [
    re.compile(r"/\.claude/"),
    re.compile(r"/node_modules/"),
    re.compile(r"^/tmp/"),
    re.compile(r"^/private/tmp/"),
]


# ── Public entry point ────────────────────────────────────────────────

def chunk_session(
    events: list[NormalizedDevEvent],
    session_id: str,
    topic_shift_event_ids: Optional[set[str]] = None,
    hard_breaks: Optional[list[int]] = None,
) -> list[SessionChunk]:
    """Split events into bounded chunks.

    topic_shift_event_ids: set of normalized event IDs classified as explicit
        topic shifts (Haiku-classified in the live pipeline; empty set on dry-run).
    hard_breaks: causalOrder values that MUST start a new chunk (e.g. sitting
        boundaries). Applied after normal split detection, before tiny-chunk merging.
    """
    if topic_shift_event_ids is None:
        topic_shift_event_ids = set()
    if hard_breaks is None:
        hard_breaks = []

    if not events:
        return []

    hard_break_orders = set(hard_breaks)

    # Short sessions
    if len(events) < SHORT_SESSION_THRESHOLD:
        if not hard_break_orders:
            return [_build_chunk(events, session_id, 0, [])]
        # Short session with hard breaks: only apply hard-break splits
        split_indices = [
            i for i in range(1, len(events))
            if events[i].causal_order in hard_break_orders
        ]
        raw_chunks = _split_at_indices(events, split_indices)
        merged = _merge_tiny_chunks(raw_chunks, hard_break_orders)
        capped = [c for chunk in merged for c in _apply_size_cap(chunk)]
        return _build_chunks_with_overlap(capped, session_id)

    # Normal path: find split points
    split_indices = _find_split_points(events, topic_shift_event_ids)

    # Apply hard breaks
    if hard_break_orders:
        existing = set(split_indices)
        for i in range(1, len(events)):
            if events[i].causal_order in hard_break_orders and i not in existing:
                split_indices.append(i)
        split_indices.sort()

    raw_chunks = _split_at_indices(events, split_indices)
    merged = _merge_tiny_chunks(raw_chunks, hard_break_orders)
    capped = [c for chunk in merged for c in _apply_size_cap(chunk)]
    return _build_chunks_with_overlap(capped, session_id)


# ── Split point detection ─────────────────────────────────────────────

def _find_split_points(
    events: list[NormalizedDevEvent],
    topic_shift_event_ids: set[str],
) -> list[int]:
    splits: list[int] = []
    last_split = 0

    for i in range(1, len(events)):
        if i - last_split < MIN_CHUNK_SIZE:
            continue

        # Priority 1: Large pause (> 5 min)
        gap = _timestamp_gap_ms(events[i - 1].timestamp, events[i].timestamp)
        if gap is not None and gap > PAUSE_THRESHOLD_MS:
            splits.append(i)
            last_split = i
            continue

        # Priority 2: File cluster shift
        if _has_file_cluster_shift(events, i):
            splits.append(i)
            last_split = i
            continue

        # Priority 3: Explicit topic shift (Haiku-classified)
        if _has_topic_shift(events[i], topic_shift_event_ids):
            splits.append(i)
            last_split = i

    return splits


def _timestamp_gap_ms(ts1: str, ts2: str) -> Optional[int]:
    ms1 = _parse_ts(ts1, None)
    ms2 = _parse_ts(ts2, None)
    if ms1 is None or ms2 is None:
        return None
    return ms2 - ms1


def _has_file_cluster_shift(events: list[NormalizedDevEvent], index: int) -> bool:
    LOOKAHEAD = 5
    LOOKBACK = 15

    forward_end = min(len(events), index + LOOKAHEAD)
    forward_files: set[str] = set()
    for j in range(index, forward_end):
        for f in _get_project_files(events[j]):
            forward_files.add(f)
    if len(forward_files) < MIN_CURRENT_FILES:
        return False

    lookback_start = max(0, index - LOOKBACK)
    previous_files: set[str] = set()
    for j in range(lookback_start, index):
        for f in _get_project_files(events[j]):
            previous_files.add(f)

    if len(previous_files) < MIN_PREVIOUS_FILES:
        return False

    forward_arr = list(forward_files)
    new_files = [f for f in forward_arr if f not in previous_files]
    return len(new_files) / len(forward_arr) > FILE_SHIFT_THRESHOLD


def _has_topic_shift(event: NormalizedDevEvent, topic_shift_event_ids: set[str]) -> bool:
    if event.category != "intent":
        return False
    return event.id in topic_shift_event_ids


# ── Merge tiny chunks ─────────────────────────────────────────────────

def _merge_tiny_chunks(
    chunks: list[list[NormalizedDevEvent]],
    hard_break_orders: set[int],
) -> list[list[NormalizedDevEvent]]:
    if len(chunks) <= 1:
        return chunks

    def starts_at_hard_break(chunk: list[NormalizedDevEvent]) -> bool:
        return bool(chunk) and chunk[0].causal_order in hard_break_orders

    result = [chunks[0]]

    for i in range(1, len(chunks)):
        is_hard_break_chunk = starts_at_hard_break(chunks[i])

        if len(chunks[i]) < MIN_CHUNK_SIZE and not is_hard_break_chunk:
            # Backward merge (safe: no hard break here)
            result[-1] = result[-1] + chunks[i]
        else:
            result.append(chunks[i])

    # If the first chunk is tiny and does NOT protect a hard-break, merge forward
    if (len(result) > 1
            and len(result[0]) < MIN_CHUNK_SIZE
            and not starts_at_hard_break(result[1])):
        result[1] = result[0] + result[1]
        result.pop(0)

    return result


# ── Size cap ──────────────────────────────────────────────────────────

def _apply_size_cap(events: list[NormalizedDevEvent]) -> list[list[NormalizedDevEvent]]:
    if len(events) <= SIZE_CAP:
        return [events]

    result: list[list[NormalizedDevEvent]] = []
    start = 0

    while start < len(events):
        remaining = len(events) - start
        if remaining <= SIZE_CAP:
            result.append(events[start:])
            break

        target = start + SIZE_CAP
        split_at = _find_natural_boundary(events, target)
        result.append(events[start:split_at])
        start = split_at

    return result


def _find_natural_boundary(events: list[NormalizedDevEvent], target: int) -> int:
    search_radius = 10
    lo = max(1, target - search_radius)
    hi = min(len(events), target + search_radius)

    # Prefer: after tool_result
    for i in range(target, lo - 1, -1):
        if events[i - 1].category == "result":
            return i
    for i in range(target + 1, hi):
        if events[i - 1].category == "result":
            return i

    # Fallback: before conversation_turn
    for i in range(target, lo - 1, -1):
        if events[i].category == "intent":
            return i
    for i in range(target + 1, hi):
        if events[i].category == "intent":
            return i

    return min(target, len(events))


# ── Chunk building ────────────────────────────────────────────────────

def _split_at_indices(
    events: list[NormalizedDevEvent],
    indices: list[int],
) -> list[list[NormalizedDevEvent]]:
    if not indices:
        return [events]

    chunks: list[list[NormalizedDevEvent]] = []
    start = 0
    for idx in indices:
        if idx > start:
            chunks.append(events[start:idx])
        start = idx
    if start < len(events):
        chunks.append(events[start:])
    return chunks


def _build_chunks_with_overlap(
    raw_chunks: list[list[NormalizedDevEvent]],
    session_id: str,
) -> list[SessionChunk]:
    result: list[SessionChunk] = []
    for i, owned_events in enumerate(raw_chunks):
        overlap_events = raw_chunks[i - 1][-OVERLAP:] if i > 0 else []
        result.append(_build_chunk(owned_events, session_id, i, overlap_events))
    return result


def _build_chunk(
    owned_events: list[NormalizedDevEvent],
    session_id: str,
    chunk_index: int,
    overlap_events: list[NormalizedDevEvent],
) -> SessionChunk:
    all_events = overlap_events + owned_events

    start_causal = owned_events[0].causal_order
    end_causal = owned_events[-1].causal_order

    return SessionChunk(
        id=f"{session_id}-chunk-{chunk_index}",
        session_id=session_id,
        chunk_index=chunk_index,
        events=all_events,
        topic_hint=_derive_topic_hint(owned_events),
        files_in_scope=_derive_files_in_scope(all_events),
        event_range=(start_causal, end_causal),
    )


# ── Topic hint & files ────────────────────────────────────────────────

def _derive_topic_hint(events: list[NormalizedDevEvent]) -> str:
    """Heuristic: most frequent file path, fallback to first user message."""
    freq: dict[str, int] = {}
    for event in events:
        for f in _get_files_affected(event):
            freq[f] = freq.get(f, 0) + 1

    if freq:
        return max(freq, key=lambda k: freq[k])

    user_event = next((e for e in events if e.category == "intent"), None)
    if user_event:
        return user_event.content.summary

    return "unknown"


def _derive_files_in_scope(events: list[NormalizedDevEvent]) -> list[str]:
    files: set[str] = set()
    for event in events:
        for f in _get_files_affected(event):
            files.add(f)
    return list(files)


def _get_files_affected(event: NormalizedDevEvent) -> list[str]:
    return event.content.files_affected or []


def _get_project_files(event: NormalizedDevEvent) -> list[str]:
    return [
        f for f in _get_files_affected(event)
        if not any(p.search(f) for p in _IGNORED_FILE_PATTERNS)
    ]
